crates/sim2h/src/connection_mgr.rs

```diff
@@ -0,0 +1,319 @@
+use crate::*;
+use std::sync::{Arc, Weak};
+
+/// incoming messages from websockets
+#[derive(Debug)]
+pub enum ConMgrEvent {
+    Disconnect(Lib3hUri, Option<Sim2hError>),
+    ReceiveData(Lib3hUri, WsFrame),
+    ConnectionCount(usize),
+}
+
+/// messages for controlling the connection manager
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug)]
+enum ConMgrCommand {
+    Connect(Lib3hUri, TcpWss),
+    SendData(Lib3hUri, WsFrame),
+    Disconnect(Lib3hUri),
+}
+
+type EvtSend = tokio::sync::mpsc::UnboundedSender<ConMgrEvent>;
+type EvtRecv = tokio::sync::mpsc::UnboundedReceiver<ConMgrEvent>;
+type CmdSend = tokio::sync::mpsc::UnboundedSender<ConMgrCommand>;
+type CmdRecv = tokio::sync::mpsc::UnboundedReceiver<ConMgrCommand>;
+
+pub type ConnectionMgrEventRecv = EvtRecv;
+
+/// internal websocket polling loop
+async fn wss_task(uri: Lib3hUri, mut wss: TcpWss, evt_send: EvtSend, mut cmd_recv: CmdRecv) {
+    let mut frame = None;
+
+    // TODO - this should be done with tokio tcp streams && selecting
+    //        for now, we're just pausing when no work happens
+
+    loop {
+        let mut did_work = false;
+
+        // first, process a batch of control commands
+        for _ in 0..100 {
+            match cmd_recv.try_recv() {
+                Ok(cmd) => {
+                    did_work = true;
+                    match cmd {
+                        ConMgrCommand::SendData(_uri, frame) => {
+                            if let Err(e) = wss.write(frame) {
+                                error!("socket write error {} {:?}", uri, e);
+                                let _ = evt_send
+                                    .send(ConMgrEvent::Disconnect(uri.clone(), Some(e.into())));
+                                // end task
+                                return;
+                            }
+                        }
+                        ConMgrCommand::Disconnect(_uri) => {
+                            debug!("disconnecting socket {}", uri);
+                            let _ = evt_send.send(ConMgrEvent::Disconnect(uri.clone(), None));
+                            // end task
+                            return;
+                        }
+                        ConMgrCommand::Connect(_, _) => unreachable!(),
+                    }
+                }
+                Err(tokio::sync::mpsc::error::TryRecvError::Empty) => break,
+                Err(tokio::sync::mpsc::error::TryRecvError::Closed) => {
+                    debug!("socket cmd channel closed {}", uri);
+                    let _ = evt_send.send(ConMgrEvent::Disconnect(uri.clone(), None));
+                    // channel broken, end task
+                    return;
+                }
+            }
+        }
+
+        // next process a batch of incoming websocket frames
+        for _ in 0..100 {
```

Comment:
Are these 0..100 special numbers?

Reply:
Just to do enough work to be useful / avoid thrashing while also not blocking other tasks - if anyone has a better idea for handling stuff like this, I'm all ears - time base solutions are heavy handed because querying the system clock is costly.